rasa_nlu/extractors/crf_entity_extractor.py

```diff
@@ -0,0 +1,241 @@
+        # type: (str, str) -> CRFEntityExtractor
+
+        if model_dir and model_name:
+            ent_tagger = pycrfsuite.Tagger()
+            ent_tagger.open(os.path.join(model_dir, 'ner', model_name))
+            config = json.load(open(os.path.join(model_dir, 'ner', 'crf_config.json'),'r'))
+
+            return CRFEntityExtractor(ent_tagger=ent_tagger, crf_features=config['crf_features'], BILOU_flag=config['BILOU_flag'])
+        else:
+            return CRFEntityExtractor()
+
+    def persist(self, model_dir):
+        # type: (str) -> dict
+        """Persist this model into the passed directory. Returns the metadata necessary to load the model again."""
+        import json
+
+        if self.f:
+            ner_dir = os.path.join(model_dir, 'ner')
+            if not os.path.exists(ner_dir):
+                os.mkdir(ner_dir)
+
+            entity_extractor_config_file = os.path.join(ner_dir, "crf_config.json")
+            entity_extractor_file = os.path.join(ner_dir, "model.crfsuite")
+            with open(entity_extractor_config_file, 'w') as f:
+                json.dump(self.config, f)
+
+            shutil.copyfileobj(self.f, open(entity_extractor_file,'w'))
+            return {
+                "entity_extractor": "ner",
+            }
+        else:
+            return {"entity_extractor": None}
+
+
+    def _sentence_to_features(self, sentence):
+        # convert a word into discrete features in self.crf_features, including word before and word after
+        sentence_features = []
+        for word_idx in xrange(len(sentence)):
+            # word before(-1), current word(0), next word(+1)  
+            prefixes = [u'-1:', u'0:', u'+1:']
+            word_features = []
+            for i in xrange(3):
+                if word_idx == len(sentence) - 1 and i == 2:
+                    word_features.append('EOS')
+                    # End Of Sentence
+                elif word_idx == 0 and i == 0:
+                    word_features.append('BOS')
+                    # Beginning Of Sentence
+                else:
+                    word = sentence[word_idx - 1 + i]
+                    prefix = prefixes[i]
+                    features = self.crf_features[i]
+                    for feature in features:
+                        # append each feature to a feature vector
+                        word_features.append(prefix + feature + ':' + self.function_dict[feature](word))
+            sentence_features.append(word_features)
+        return sentence_features
+
+    def _sentence_to_labels(self, sentence):
+        labels = []
+        for word in sentence:
+            labels.append(word[2])
+        return labels
+
+    def _from_json_to_crf(self, json_eg):
+            #takes the json examples and switches them to a format which crfsuite likes
+            doc = self.nlp(json_eg[0])
+            entity_offsets = json_eg[1]
+            gold = GoldParse(doc, entities=entity_offsets)
```

Comment:
that and converting the (start, end, entity) format entity_offsets to labelled words